src/pint/models/model_builder.py

```diff
@@ -26,276 +26,296 @@ class UnknownBinaryModel(ValueError):
+        primary_key: str
+            The primary compare key(e.g., `params` or `component_special_params`)
 
+        Return
+        ------
+        A sorted tuple list based on how many overlap parameters.
         """
-        sorted_components = []
-        for cat in self.get_all_categories():
-            # FIXME, I am not sure adding orders here is a good idea.
-            if cat not in category_order:
-                category_order.append(cat)
-        for co in category_order:
-            if co not in self.select_comp:
-                continue
-            cp = self.select_comp[co]
-            sorted_components.append(cp)
-        return sorted_components
-
-    def search_prefix_param(self, paramList, model, prefix_type):
-        """Check if the Unrecognized parameter has prefix parameter"""
-        prefixs = {}
-        prefix_inModel = model.get_params_of_type_top(prefix_type)
-        for pn in prefix_inModel:
-            par = getattr(model, pn)
-            prefixs[par.prefix] = []
-            for p in paramList:
-                try:
-                    pre, idxstr, idxV = split_prefixed_name(p)
-                    if pre in [par.prefix] + par.prefix_aliases:
-                        prefixs[par.prefix].append(p)
-                except ValueError:  # FIXME: is this meant to catch KeyErrors?
-                    continue
+        param_match = {}
+        for cp in components:
+            check_param = getattr(self.components[cp], primary_key)
+            overlap = list(set(params) & set(check_param))
+            param_match[cp] = len(overlap)
+        # Selected the best overlap
+        # Get the max overlap
+        rank = sorted(list(param_match.items()), key=lambda tup: tup[1])
+        return rank
+
+    def choose_model(self, parfile):
+        """ Choose the model components based on the parfile.
+
+        Parameter
+        ---------
+        parfile: str
+            parfile name and StringIO of parfile.
 
-        return prefixs
+        Return
+        ------
+        List of selected components and a dictionary of conflict components.
 
-    def build_model(self, parfile=None, name=""):
-        """Read parfile using the model_instance attribute. Throws error if
-           mismatched coordinate systems detected.
-        Parameters
-        ---------
-        name: str, optional
-            The name for the timing model
-        parfile : str optional
-            The parfile name
+        Note
+        ----
+        The selection algorithm:
+        1. Look at the BINARY in the par file and catche the indicated binary model
+        2. Translate para file parameters to the pint parameter name
+        3. Get all the components which holds the parameters
+        4. Check if there is any conflict in the component(i.e., mulitple
+           components in one category)
+          4.1 Find the overlap between component special special parameter and
+              in parfile component.
```

Comment:
I'm not sure I understand this resolution mechanism. This algorithm needs to be something we can explain to users - what Components end up in their model.

Reply:
This is changed.